spec/wordcounter_spec.rb

```diff
@@ -0,0 +1,164 @@
+require 'rspec'
+require 'tempfile'
+require_relative '../lib/wordcounter'
+
+PROJECT_ROOT = File.join(File.dirname(__FILE__), '..')
+
+RSpec::Matchers.define :not_match do |regex|
+  match do |actual|
+    actual !~ regex
+  end
+
+  failure_message do |actual|
+    "expected `#{actual}` to not match #{regex}"
+  end
+end
+
+RSpec.shared_examples 'word counter' do |method, *args|
+  context "##{method.to_s}" do
+    subject(:counter) { WordCounter.new }
+
+    it 'returns a histogram' do
+      expect(counter.send(method, *args)).to be_a(WordHistogram)
+    end
+
+    it 'ignores punctuation' do
+      histogram = counter.send(method, *args)
+      expect(histogram.keys).to all(not_match /(\w?[.,\/#!?$%^&*;:{}=`+~()<>"])|(\w[-_])|([-_]\w)/)
+    end
+
+    it 'ignores case, forcing to lowercase' do
+      histogram = counter.send(method, *args)
+      keys = histogram.keys
+      expect(keys).to match_array(keys.map &:downcase)
+    end
+
+    it 'is able to not ignore case' do
+      histogram = counter.send(method, *args, ignorecase: false)
+      keys = histogram.keys
+
+      expect(keys).not_to match_array(keys.map &:downcase)
+    end
+
+    it 'only includes word characters' do
+      histogram = counter.send(method, *args)
+      expect(histogram.keys).to all(match /^[\w'\-_]+$/)
+    end
+  end
+end
+
+describe 'WordHistogram' do
+  let(:hist) { WordHistogram.new }
+
+  it 'defaults to a word frequency of 0' do
+    expect(hist['a']).to eq 0
+  end
+
+  before { hist[:a] += 1 }
```

Comment:
In rspec, nesting matters, line-order doesn't.  This `before` block is going to run before every test in the `describe 'WordHistogram'` block.